presto-cli/src/main/java/com/facebook/presto/cli/QueryPreprocessor.java

```diff
@@ -0,0 +1,162 @@
+    {
+        // execute the process in a child thread so we can better handle interruption and timeouts
+        AtomicReference<Process> processReference = new AtomicReference<>();
+        SettableFuture<String> resultFuture = SettableFuture.create();
+        Thread thread = new Thread(() -> {
+            try {
+                ProcessBuilder processBuilder = new ProcessBuilder(preprocessorCommand);
+                processBuilder.redirectErrorStream(true);
+                processBuilder.environment().put(ENV_PRESTO_CATALOG, catalog.orElse(""));
+                processBuilder.environment().put(ENV_PRESTO_SCHEMA, schema.orElse(""));
+
+                Process process = processBuilder.start();
+                try {
+                    // write query to process standard out
+                    OutputStream outputStream = process.getOutputStream();
+                    outputStream.write(query.getBytes(UTF_8));
+                    process.getOutputStream().flush();
+                    process.getOutputStream().close();
+
+                    // read response
+                    String result = CharStreams.toString(new InputStreamReader(process.getInputStream(), UTF_8));
+
+                    // wait for process to finish
+                    int exitCode = process.waitFor();
+
+                    // check we got a valid exit code
+                    if (exitCode == 0) {
+                        resultFuture.set(result);
+                    }
+                    else {
+                        resultFuture.setException(new QueryPreprocessorException("Query preprocessor failed:\n\n" + result));
+                    }
+                }
+                finally {
+                    process.destroyForcibly();
+                }
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                resultFuture.setException(new QueryPreprocessorException("Interrupted while preprocessing query"));
+            }
+            catch (Throwable e) {
+                resultFuture.setException(new QueryPreprocessorException("Error preprocessing query: " + e.getMessage(), e));
+            }
+        });
+        thread.setName("Query preprocessor");
+        thread.start();
+
+        try {
+            return resultFuture.get(timeout.toMillis(), MILLISECONDS);
+        }
+        catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new QueryPreprocessorException("Interrupted while preprocessing query");
+        }
+        catch (ExecutionException e) {
+            Throwable cause = e.getCause();
+            propagateIfPossible(cause, QueryPreprocessorException.class);
+            throw new QueryPreprocessorException("Error preprocessing query: " + cause.getMessage(), cause);
+        }
+        catch (TimeoutException e) {
+            throw new QueryPreprocessorException("Timed out waiting for query preprocessor");
```

Comment:
Since the timeout value is not explicit (taken from env), quoting it here might turn useful